[keyvault] administration package readme (#13489)
Azure/azure-sdk-for-python: sdk/keyvault/azure-keyvault-administration/README.md

> # Azure KeyVault Administration client library for .NET

Reviewer: Was Python supposed to be changed to .NET?

Author: oh jeez, thanks for catching all of my embarrassing copy-past errors

Reviewer: that's what I'm here for ✨

role_assignment = client.get_role_assignment(role_scope, role_assignment.name)
print(role_assignment.name)
print(role_assignment.principal_id)
print(role_assignment.role_definition_id)

# finally, we delete this role assignment
role_assignment = client.delete_role_assignment(role_scope, role_assignment.name)
print(role_assignment.name)
print(role_assignment.principal_id)
print(role_assignment.role_definition_id)
```

### Perform a full key backup
Back up your entire collection of keys. The backing store for full key backups is a blob storage container using Shared Access Signature authentication.

For more details on creating a SAS token using the `BlobServiceClient`, see the sample [here](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/storage/azure-storage-blob/samples/blob_samples_authentication.py#L105).
Alternatively, it is possible to [generate a SAS token in Storage Explorer](https://docs.microsoft.com/en-us/azure/vs-azure-tools-storage-manage-with-storage-explorer?tabs=windows#generate-a-shared-access-signature-in-storage-explorer)

```python
from azure.identity import DefaultAzureCredential
from azure.keyvault.administration import KeyVaultBackupClient

> from azure.core.exceptions import ResourceNotFoundError

Reviewer 2: ```suggestion

```

blob_storage_uri = "<your-blob-storage-uri>"  # the URI to your storage account. Should contain the name of the specific container
sas_token = "<your-sas-token>"  # replace with the sas token to your storage account. See this snippet's description on help to retrieve

# performing a full key backup is a long-running operation. Calling `result()` on the poller will wait
# until the backup is completed, then return an object representing the backup operation.
backup_operation = client.begin_full_backup(blob_storage_uri, sas_token).result()

# this is the URI of the Azure blob storage container which contains the backup
azure_storage_blob_container_uri = backup_operation.azure_storage_blob_container_uri

print(backup_operation.status)
print(backup_operation.job_id)
print(azure_storage_blob_container_uri)
```

### Perform a full key restore
Restore your entire collection of keys from a backup. The data source for a full key restore is a storage blob accessed using Shared Access Signature authentication.
You will also need the `azure_storage_blob_container_uri` from the [above snippet](#perform-a-full-key-backup).

> For more details on creating a SAS token using the `BlobServiceClient`, see the sample [here](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/storage/azure-storage-blob/samples/blob_samples_authentication.py#L105).

Author: linked to the specific snippet that demonstrates this, since I think that the sample should be constant from now on. Let me know if you think this should be linked to the sample directory instead for more resilience.

Alternatively, it is possible to [generate a SAS token in Storage Explorer](https://docs.microsoft.com/en-us/azure/vs-azure-tools-storage-manage-with-storage-explorer?tabs=windows#generate-a-shared-access-signature-in-storage-explorer)

```python
from azure.identity import DefaultAzureCredential
from azure.keyvault.administration import KeyVaultBackupClient

> from azure.core.exceptions import ResourceNotFoundError

Reviewer 2: ```suggestion
```

# Azure KeyVault Administration client library for .NET
Azure Key Vault helps solve the following problems:
- Vault administration (this library) - full backup / restore, and key-level

> role-based access control (RBAC) of vaults.

Reviewer 2: These bullets render as expected. Single newlines are fine, it's a blank line that ends the bullet.

- Cryptographic key management ([azure-keyvault-keys](https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/keyvault/azure-keyvault-keys)) - create, store, and control

> access to the keys used to encrypt your data

Reviewer 3: A few more below as well. If using code, switch to the preview window to render it and see if it looks right. Note that it's rendering may differ from GitHub's, though they are usually pretty close.

> access to the keys used to encrypt your data

Author: It's rendering well, thanks @heaths for double checking

- Secrets management
([azure-keyvault-secrets](https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/keyvault/azure-keyvault-secrets)) -
securely store and control access to tokens, passwords, certificates, API keys,
and other secrets
- Certificate management
([azure-keyvault-certificates](https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/keyvault/azure-keyvault-certificates)) -
create, manage, and deploy public and private SSL/TLS certificates

## Getting started
### Install packages
Install [azure-keyvault-administration][pypi_package_administration] and
[azure-identity][azure_identity_pypi] with [pip][pip]:
```Bash

> pip install azure-keyvault-administration azure-identity

Reviewer 3: Don't you need to also pass `--pre`?

Author: We never really did that with our other beta backages, @chlowell are you good with keeping it without the flag?

Reviewer 2: Yes, `--pre` isn't required when a package has only beta versions.

```
[azure-identity][azure_identity] is used for Azure Active Directory
authentication as demonstrated below.

### Prerequisites
* An [Azure subscription][azure_sub]
* Python 2.7, 3.5.3, or later
* A Key Vault. If you need to create one, See the final two steps in the next section for details on creating the Key Vault with the Azure CLI.

### Authenticate the client
This document demonstrates using [DefaultAzureCredential][default_cred_ref]
to authenticate as a service principal. However, [KeyVaultAccessControlClient][rbac_client_docs]
accepts any [azure-identity][azure_identity] credential. See the
[azure-identity][azure_identity] documentation for more information about other
credentials.

#### Create/Get credentials
This [Azure Cloud Shell][azure_cloud_shell] snippet shows how to create a
new service principal. Before using it, replace "your-application-name" with
a more appropriate name for your service principal.

* Create a service principal:
    ```Bash

>     az ad sp create-for-rbac --name http://my-application --skip-assignment

Reviewer: The text before this refers to "your-application-name" but the code snipped uses "my-application". Minor, but should probably be the same

Author: great point, changing all to "your-application-name"

    ```

#### Create a client
Once the **AZURE_CLIENT_ID**, **AZURE_CLIENT_SECRET** and
**AZURE_TENANT_ID** environment variables are set,
[DefaultAzureCredential][default_cred_ref] will be able to authenticate the
[KeyVaultAccessControlClient][rbac_client_docs].

Constructing the client also requires your vault's URL, which you can
get from the Azure CLI or the Azure Portal. In the Azure Portal, this URL is
the vault's "DNS Name".

```python
from azure.identity import DefaultAzureCredential
from azure.keyvault.administration import KeyVaultAccessControlClient

credential = DefaultAzureCredential()

client = KeyVaultAccessControlClient(vault_url="https://my-key-vault.vault.azure.net/", credential=credential)
```
## Key concepts

### Role Definition
A role definition defines the operations that can be performed, such as read, write, and delete. It can also define the operations that are excluded from allowed operations.

A role definition is specified as part of a role assignment.

### Role Assignment.
A role assignment is the association of a role definition to a service principal. They can be created, listed, fetched individually, and deleted.

### KeyVaultAccessControlClient

> A `KeyVaultAccessControlClient` provides both synchronous and asynchronous operations allowing for management of role definitions and role assignments.

Reviewer 2: To avoid suggesting the same object has sync and async methods:
```suggestion
A `KeyVaultAccessControlClient` manages role definitions and role assignments.
```

## Examples
This section conntains code snippets covering common tasks:
* [List the role definitions](#list-the-role-definitions "List the role definitions")
* [Create, Get, and Delete a role assignment](#create-get-and-delete-a-role-assignment "Create, Get, and Delete a role assignment")

### List the role definitions
List the role definitions available for assignment.

```python
from azure.identity import DefaultAzureCredential
from azure.keyvault.administration import KeyVaultAccessControlClient

credential = DefaultAzureCredential()

client = KeyVaultAccessControlClient(vault_url="https://my-key-vault.vault.azure.net/", credential=credential)

> 
> role_definitions = client.list_role_definitions(role_scope="/")  # this is the global scope

Reviewer 2: ```suggestion
# "/" is the global scope. This will list all role definitions available for assignment:
role_definitions = client.list_role_definitions(role_scope="/")
```

for role_definition in role_definitions:
    print(role_definition.id)
    print(role_definition.role_name)
    print(role_definition.description)
```

### Create, Get, and Delete a role assignment
Assign a role to a service principal. This will require a role definition id from the list retrieved in the [above snippet](#list-the-role-definitions) and the principal object id retrieved in the [Create/Get credentials](#create/get-credentials)

```python
from azure.identity import DefaultAzureCredential
from azure.keyvault.administration import KeyVaultAccessControlClient

credential = DefaultAzureCredential()

client = KeyVaultAccessControlClient(vault_url="https://my-key-vault.vault.azure.net/", credential=credential)

role_scope = "/"  # setting the scope to the global scope

> role_assignment_name = "<role-assignment-name>"  # must be a UUID

Reviewer 2: ```suggestion
role_assignment_name = uuid.uuid4()
```
(and `import uuid` above). Note to self: would be nice to do this by default.

Author: Yeah that would be nice, I think .net is doing this. Another nice thing to add would be an enum for the scope like .net [here](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultRoleScope.cs). I can pick up both of these fixes.

Author: I think I was wrong about .net doing this, it just didn't show up in the readme, but it appears it shows up in the actual client @heaths for confirmation

Reviewer 3: Yes, in .NET we do auto-create GUIDs if not passed in.

> role_definition = "<role-definition>"  # Replace <role-definition-id> with a role definition from the definitions returned from the previous example

Reviewer 2: ```suggestion
role_definition_id = "<role-definition-id>"  # Replace <role-definition-id> with the id of a definition returned from the previous example
```

principal_id = "<your-service-principal-object-id>"

# first, let's create the role assignment

> role_assignment = client.create_role_assignment(role_scope, role_assignment_name, role_definition.id, principal_id)

Reviewer 2: ```suggestion
role_assignment = client.create_role_assignment(role_scope, role_assignment_name, role_definition_id, principal_id)
```

    client.get_role_assignment("/", "which-does-not-exist")
except ResourceNotFoundError as e:
    print(e.message)
```

## Next steps

Content forthcoming

## Contributing

This project welcomes contributions and suggestions.  Most contributions require
you to agree to a Contributor License Agreement (CLA) declaring that you have
the right to, and actually do, grant us the rights to use your contribution. For
details, visit [cla.microsoft.com][cla].

This project has adopted the [Microsoft Open Source Code of Conduct][coc].
For more information see the [Code of Conduct FAQ][coc_faq]
or contact [opencode@microsoft.com][coc_contact] with any
additional questions or comments.

<!-- LINKS -->
[azure_cloud_shell]: https://shell.azure.com/bash
[azure_core_exceptions]: https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/core/azure-core#azure-core-library-exceptions
[azure_identity]: https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/identity/azure-identity
[azure_identity_pypi]: https://pypi.org/project/azure-identity/
[azure_sub]: https://azure.microsoft.com/free/
[default_cred_ref]: https://aka.ms/azsdk/python/identity/docs#azure.identity.DefaultAzureCredential

> [code_of_conduct]: https://opensource.microsoft.com/codeofconduct/

Reviewer: Does [code_of_conduct] need to be changed to [coc] for this link to work? Also, the preview window suggests that links for [cla], [coc_contact] and [coc_faq] need to be added.

Author: yep, thanks! i'm going to not be lazy and check all links now too

see the Code of Conduct FAQ or contact opencode@microsoft.com with any
additional questions or comments.

<!-- LINKS -->
[azure_cloud_shell]: https://shell.azure.com/bash
[azure_core_exceptions]: https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/core/azure-core#azure-core-library-exceptions
[azure_identity]: https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/identity/azure-identity
[azure_identity_pypi]: https://pypi.org/project/azure-identity/
[azure_sub]: https://azure.microsoft.com/free/
[default_cred_ref]: https://aka.ms/azsdk/python/identity/docs#azure.identity.DefaultAzureCredential
[code_of_conduct]: https://opensource.microsoft.com/codeofconduct/
[keyvault_docs]: https://docs.microsoft.com/azure/key-vault/
[pip]: https://pypi.org/project/pip/
[pypi_package_administration]: https://aka.ms/azsdk/python/keyvault-administration/pypi
[reference_docs]: https://aka.ms/azsdk/python/keyvault-administration/docs

> ![Impressions](https://azure-sdk-impressions.azurewebsites.net/api/impressions/azure-sdk-for-net%2Fsdk%2Ftables%2FAzure.Data.Tables%2FREADME.png)

Reviewer 2: ```suggestion
![Impressions](https://azure-sdk-impressions.azurewebsites.net/api/impressions/azure-sdk-for-python%2Fsdk%2Fkeyvault%2Fazure-keyvault-administration%2FREADME.png)
```
